File: music-application/src/main/java/com/music/sale/application/auth/service/AuthService.java
package com.music.sale.application.auth.service;

import com.music.sale.application.auth.dto.RefreshTokenCommand;
import com.music.sale.application.auth.dto.SocialLoginCommand;
import com.music.sale.application.auth.dto.TokenResponse;
import com.music.sale.application.auth.port.in.AuthUseCase;
import com.music.sale.application.auth.port.out.JwtTokenPort;
import com.music.sale.application.auth.port.out.RefreshTokenPort;
import com.music.sale.application.user.dto.input.CreateSocialUserInput;
import com.music.sale.application.user.port.inport.UserCommandUseCase;
import com.music.sale.application.user.port.inport.UserQueryUseCase;
import com.music.sale.common.BusinessException;
import com.music.sale.common.ErrorDefinition;
import com.music.sale.domain.user.User;
import java.time.Duration;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional
public class AuthService implements AuthUseCase {

  private static final Duration REFRESH_TOKEN_RENEWAL_THRESHOLD = Duration.ofDays(3);

  private final UserQueryUseCase userQueryUseCase;
  private final UserCommandUseCase userCommandUseCase;
  private final JwtTokenPort jwtTokenPort;
  private final RefreshTokenPort refreshTokenPort;

  @Override
  public TokenResponse loginWithSocialAccount(
      SocialLoginCommand command, String ip, String deviceInfo) {
    User user =
        userQueryUseCase
            .findBySocialAccount(command.provider(), command.providerId())
            .orElseGet(() -> createNewSocialUser(command));

    user.validateActive();

    String accessToken = jwtTokenPort.generateAccessToken(user);
    String refreshToken = jwtTokenPort.generateRefreshToken(user);

    refreshTokenPort.saveRefreshToken(refreshToken, user.getId(), ip, deviceInfo);

    return TokenResponse.of(accessToken, refreshToken, jwtTokenPort.getAccessTokenValidity());
  }

  @Override
  public TokenResponse renewAccessToken(RefreshTokenCommand command) {
    Long userId = findUserIdByRefreshToken(command.refreshToken());
    User user = findActiveUserById(userId);

    String newAccessToken = jwtTokenPort.generateAccessToken(user);

    if (shouldRenewRefreshToken(command.refreshToken())) {
      return renewBothTokens(user, command, newAccessToken);
    }

    return TokenResponse.of(
        newAccessToken, command.refreshToken(), jwtTokenPort.getAccessTokenValidity());
  }

  @Override
  public void logout(String refreshToken) {
    refreshTokenPort.deleteRefreshToken(refreshToken);
  }

  private User createNewSocialUser(SocialLoginCommand command) {
    CreateSocialUserInput input =
        CreateSocialUserInput.of(
            command.provider(),
            command.providerId(),
            command.email(),
            command.nickname(),
            command.profileImageUrl());

    return userCommandUseCase.createSocialUser(input);
  }

  private Long findUserIdByRefreshToken(String refreshToken) {
    return refreshTokenPort
        .getUserIdByRefreshToken(refreshToken)
        .orElseThrow(() -> new BusinessException(AuthErrorCode.INVALID_REFRESH_TOKEN));
  }

  private User findActiveUserById(Long userId) {
    User user =
        userQueryUseCase
            .findById(userId)
            .orElseThrow(() -> new BusinessException(AuthErrorCode.USER_NOT_FOUND));

    user.validateActive();
    return user;
  }

  private boolean shouldRenewRefreshToken(String refreshToken) {
    Long remainingTimeMillis = refreshTokenPort.getRefreshTokenTTL(refreshToken);
    return remainingTimeMillis != null
        && remainingTimeMillis < REFRESH_TOKEN_RENEWAL_THRESHOLD.toMillis();
  }

  private TokenResponse renewBothTokens(
      User user, RefreshTokenCommand command, String newAccessToken) {
    String newRefreshToken = jwtTokenPort.generateRefreshToken(user);

    refreshTokenPort.deleteRefreshToken(command.refreshToken());
    refreshTokenPort.saveRefreshToken(
        newRefreshToken, user.getId(), command.ip(), command.deviceInfo());

    return TokenResponse.of(newAccessToken, newRefreshToken, jwtTokenPort.getAccessTokenValidity());
  }

  private enum AuthErrorCode implements ErrorDefinition {
    USER_NOT_FOUND("error.USER_NOT_FOUND"),
    INVALID_REFRESH_TOKEN("error.INVALID_REFRESH_TOKEN");

    private final String key;

    AuthErrorCode(String key) {
      this.key = key;
    }

    @Override
    public String getKey() {
      return key;
    }
  }
}

File: music-application/src/main/kotlin/com/music/sale/application/auth/service/AuthService.kt
// Copyright (C) 2024 Your Name or Company
package com.music.sale.application.auth.service

import com.music.sale.application.auth.dto.AuthResponse
import com.music.sale.application.auth.dto.LoginRequest
import com.music.sale.application.auth.dto.RegisterRequest
import com.music.sale.application.user.port.outport.UserPort
import com.music.sale.domain.user.User
import com.music.sale.domain.user.enum.UserRole
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
open class AuthService(
        private val userPort: UserPort,
        private val jwtService: JwtService,
        private val passwordEncoder: PasswordEncoder,
) {
    fun login(request: LoginRequest): AuthResponse {
        val user =
                userPort.findByEmail(request.email)
                        ?: throw IllegalArgumentException("Invalid email or password")

        // TODO: 실제 비밀번호 검증 로직 구현 필요
        // 현재는 임시로 이메일만 확인
        if (user.email?.value != request.email) {
            throw IllegalArgumentException("Invalid email or password")
        }

        val token =
                jwtService.generateToken(
                        userId = user.id ?: 0L,
                        email = user.email?.value ?: "",
                        role = user.role ?: UserRole.USER,
                )

        return AuthResponse(
                token = token,
                userId = user.id ?: 0L,
                email = user.email?.value ?: "",
                name = user.name?.value ?: "",
                role = user.role ?: UserRole.USER,
        )
    }

    fun register(request: RegisterRequest): AuthResponse {
        // 이메일 중복 확인
        val existingUser = userPort.findByEmail(request.email)
        if (existingUser != null) {
            throw IllegalArgumentException("Email already exists")
        }

        // 새 사용자 생성
        val newUser =
                User(
                        id = null,
                        email = User.Email(request.email),
                        name = User.Name(request.name),
                        role = request.role,
                        provider = null,
                        providerId = null,
                )

        val savedUser = userPort.saveEmail(newUser, request.password)

        val token =
                jwtService.generateToken(
                        userId = savedUser.id ?: 0L,
                        email = savedUser.email?.value ?: "",
                        role = savedUser.role ?: UserRole.USER,
                )

        return AuthResponse(
                token = token,
                userId = savedUser.id ?: 0L,
                email = savedUser.email?.value ?: "",
                name = savedUser.name?.value ?: "",
                role = savedUser.role ?: UserRole.USER,
        )
    }
}
